### PR TITLE
Add examples for `PlutusV2` reference scripts and reference inputs

### PR DESCRIPTION
**Pull request by errfrom**
Partially closes https://github.com/Plutonomicon/cardano-transaction-lib/issues/922

### Changes
1. **[Breaking change]** Updated `Plutus.Utxo` to be a type synonym for `Map TransactionInput TransactionOutputWithRefScript`, so the return type of `utxosAt` has changed. This was motivated by the fact that we need to have access to full scripts of unspent outputs in `ScriptLookups`, and not just their hashes.

2. Added new constraints for creating outputs with reference scripts: `mustPayWithScriptRefToPubKey`, `mustPayWithScriptRefToPubKeyAddress`, `mustPayWithScriptRefToScript`, `mustPayWithDatumAndScriptRefToPubKey`, `mustPayWithDatumAndScriptRefToPubKeyAddress`.

3. Added new constraints for using reference validators and minting policies: `mustSpendScriptOutputUsingScriptRef`, `mustMintCurrencyUsingScriptRef`, `mustMintCurrencyWithRedeemerUsingScriptRef`.

4. Added new constraint for attaching a reference input to a transaction: `mustReferenceOutput`.

5. Changed the type of `referenceInputs` field of `TxBody` from `Maybe (Array TransactionInput)` to `Set TransactionInput` to comply with [Babbage CDDL](https://github.com/input-output-hk/cardano-ledger/blob/master/eras/babbage/test-suite/cddl-files/babbage.cddl) and to prevent attaching the same reference input twice.

6. Consider reference scripts to determine the language set required for script integrity hash calculation ([Babbage Specification](https://hydra.iohk.io/build/17586760/download/1/babbage-changes.pdf), Figure 3).

### Pre-review checklist

- [x] All code has been formatted using our config (`make format`)
- [x] Any new API feature includes a full example or integration test run with Plutip on CI
- [x] The changelog has been updated under the `## Unreleased` header, using the appropriate sub-headings (`### Added`, `### Removed`, `### Fixed`), and the links to the appropriate issues/PRs have been included

### CODE REVIEW

File: src/Types/TxConstraints.purs [hunk 1]
```diff
@@ -186,6 +196,56 @@ mustBeSignedBy = singleton <<< MustBeSignedBy
 mustIncludeDatum :: forall (i :: Type) (o :: Type). Datum -> TxConstraints i o
 mustIncludeDatum = singleton <<< MustIncludeDatum
 
+-- | Require the transaction to include a reference input.
+mustIncludeReferenceInput
```
Comment (ngua) [inline src/Types/TxConstraints.purs:200]: I thought we were going to wait to add the new constraints until after the v2.0.0 release, but I'm reconsidering that now
- you've already implemented these
- the changes are largely backwards-compatible and unobtrusive
- we have a growing number of projects who want to use the new features (at least three for sure)

So I think we should keep the new constraints. We just need to deal with `Datum` arguments to other ones (not related to this PR)

File: src/Types/TxConstraints.purs [hunk 2]
```diff
@@ -186,6 +212,57 @@ mustBeSignedBy = singleton <<< MustBeSignedBy
...
+   . PaymentPubKeyHash
+  -> StakePubKeyHash
+  -> Value
+  -> TxConstraints i o
+mustPayToPubKeyAddress pkh skh =
+  singleton <<< MustPayToPubKeyAddress pkh (Just skh) Nothing Nothing
+
+-- | Lock the value and datum with a payment public key hash and (optionally) a
+-- | stake public key hash.
+mustPayWithDatumToPubKeyAddress
```
Comment (klntsky) [inline src/Types/TxConstraints.purs:234]: `mustPayToPubKeyAddressWithDatum`

File: src/Types/TxConstraints.purs [hunk 3]
```diff
@@ -186,6 +212,57 @@ mustBeSignedBy = singleton <<< MustBeSignedBy
...
+  :: forall (i :: Type) (o :: Type)
+   . PaymentPubKeyHash
+  -> StakePubKeyHash
+  -> Datum
+  -> Value
+  -> TxConstraints i o
+mustPayWithDatumToPubKeyAddress pkh skh datum =
+  singleton <<< MustPayToPubKeyAddress pkh (Just skh) (Just datum) Nothing
+
+mustPayWithScriptRefToPubKeyAddress
```
Comment (klntsky) [inline src/Types/TxConstraints.purs:244]: `mustPayToPubKeyAddressWithScriptRef`

File: src/Types/TxConstraints.purs [hunk 4]
```diff
@@ -186,6 +212,57 @@ mustBeSignedBy = singleton <<< MustBeSignedBy
...
+  :: forall (i :: Type) (o :: Type)
+   . PaymentPubKeyHash
+  -> StakePubKeyHash
+  -> ScriptRef
+  -> Value
+  -> TxConstraints i o
+mustPayWithScriptRefToPubKeyAddress pkh skh scriptRef =
+  singleton <<< MustPayToPubKeyAddress pkh (Just skh) Nothing (Just scriptRef)
+
+mustPayWithDatumAndScriptRefToPubKeyAddress
```
Comment (klntsky) [inline src/Types/TxConstraints.purs:254]: `mustPayToPubKeyAddressWithDatumAndScriptRef`

File: src/Types/TxConstraints.purs [hunk 5]
```diff
@@ -196,40 +273,37 @@ mustPayToPubKey
...
   -> TxConstraints i o
-mustPayToPubKey pkh = singleton <<< MustPayToPubKeyAddress pkh Nothing Nothing
+mustPayToPubKey pkh =
+  singleton <<< MustPayToPubKeyAddress pkh Nothing Nothing Nothing
 
--- | Lock the value with a payment public key hash and (optionally) a stake
--- | public key hash.
-mustPayToPubKeyAddress
+-- | Lock the value and datum with a payment public key hash
+mustPayWithDatumToPubKey
```
Comment (klntsky) [inline src/Types/TxConstraints.purs:280]: `mustPayToPubKeyWithDatum`

File: src/Types/TxConstraints.purs [hunk 6]
```diff
@@ -196,40 +273,37 @@ mustPayToPubKey
...
   -> Value
   -> TxConstraints i o
-mustPayToPubKeyAddress pkh skh =
-  singleton <<< MustPayToPubKeyAddress pkh (Just skh) Nothing
+mustPayWithDatumToPubKey pkh datum =
+  singleton <<< MustPayToPubKeyAddress pkh Nothing (Just datum) Nothing
 
--- | Lock the value and datum with a payment public key hash
-mustPayWithDatumToPubKey
+mustPayWithScriptRefToPubKey
```
Comment (klntsky) [inline src/Types/TxConstraints.purs:289]: `mustPayToPubKeyWithScriptRef`

File: src/Types/TxConstraints.purs [hunk 7]
```diff
@@ -196,40 +273,37 @@ mustPayToPubKey
...
-mustPayWithDatumToPubKey pkh datum =
-  singleton <<< MustPayToPubKeyAddress pkh Nothing (Just datum)
+mustPayWithScriptRefToPubKey pkh scriptRef =
+  singleton <<< MustPayToPubKeyAddress pkh Nothing Nothing (Just scriptRef)
 
--- | Lock the value and datum with a payment public key hash and (optionally) a
--- | stake public key hash.
-mustPayWithDatumToPubKeyAddress
-  :: forall i o
+mustPayWithDatumAndScriptRefToPubKey
```
Comment (klntsky) [inline src/Types/TxConstraints.purs:298]: `mustPayToPubKeyWithDatumAndScriptRef`

File: src/Types/TxConstraints.purs [hunk 8]
```diff
@@ -243,7 +317,18 @@ mustPayToScript
   -> Value
   -> TxConstraints i o
 mustPayToScript vh dt vl =
-  singleton (MustPayToScript vh dt vl)
+  singleton (MustPayToScript vh dt Nothing vl)
+    <> singleton (MustIncludeDatum dt)
+
+mustPayWithScriptRefToScript
```
Comment (klntsky) [inline src/Types/TxConstraints.purs:323]: `mustPayToScriptWithScriptRef`

**Review by klntsky — CHANGES_REQUESTED** [review]
This naming sounds better. Also, most likely user will search for functions using a prefix like `mustPayTo...` with docs-search, that can only look up prefixes.

**Review by klntsky — APPROVED** [review]

**Comment by ngua** [comment]
I'm going to build the checks locally since Hydra is down. If it works I'll merge this

**Review by ngua — APPROVED** [review]
The checks work locally, merging

**Pull request merged**